### ISSUE TITLE
`sequence $ repeat ...` should fire a warning

### ISSUE THREAD

**Issue by bitemyapp** [ndmitchell/hlint#407]
Just saw somebody bump into this, didn't know `sequence` forced the whole shebang.

Possibly other things that generate infinite lists should get a rule too.

Original code questioner posted:

```haskell
-- | natural selection, probablistic to fitness compared against group
-- indexes in [Individual] correspond to the same index in [Fitness]
probabilisticFitnessSelection :: MonadIO m => [Individual] -> [Float] -> m [Individual]
probabilisticFitnessSelection individuals fitnesses = do
  let fitSum = sum fitnesses -- use for normalizing: x / fitSum
      accumProbs = scanl1 (+) $ map (\x -> x / fitSum) fitnesses -- stack probabilities together
      zipped = zip accumProbs individuals
  newProbs <- sequence $ repeat (liftIO randomIO :: MonadIO m => m Float)
  pure $ [snd . head $ filter (\(prob, ind) -> target < prob) zipped
         | target <- newProbs] -- select individuals whose accumulated fitness probability is more than the target
```

**Comment by ndmitchell** [comment]
I think it depends entirely on the Monad. In `IO`, `sequence $ repeat $ print 1` is equivalent to `forever $ print 1`. In `Maybe`, `sequence $ repeat Nothing == Nothing`, while `sequence $ repeat $ Just 1` loops. Not sure there's any general rule here, without knowing the specific monad, which HLint doesn't. There are a few rules about infinite lists and repeat, but not sure I can think of anything else from the above example.

**Comment by ndmitchell** [comment]
I don't think there's anything monad-agnostic here so closing this as not possible for HLint to do syntactically.
